### PR TITLE
More efficient Value.geq and leq

### PR DESCRIPTION
**Pull request by zliu41**

### CODE REVIEW

File: plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs [hunk 1]
```diff
@@ -316,40 +316,22 @@ flattenValue v = goOuter [] (Map.toList $ getValue v)
...
-    in checkPred unThese l r
-
 {-# INLINABLE geq #-}
 -- | Check whether one 'Value' is greater than or equal to another. See 'Value' for an explanation
 -- of how operations on 'Value's work.
 geq :: Value -> Value -> Bool
--- If both are zero then checkBinRel will be vacuously true, but this is fine.
-geq = checkBinRel (>=)
+geq l (Value r) =
+  all
```
Comment (michaelpj) [inline plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs:324]: Maybe define `valueEntries :: Value -> [(CurrencyId, TokenId, Integer)]` function? Useful here and probably elsewhere.

Comment (zliu41) [reply plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs:324]: It already exists: `flattenValue`, but flattening both values before checking `geq` is more expensive than this.

Comment (michaelpj) [reply plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs:324]: Aren't you pretty much exactly flattening the second value? I guess you don't construct an intermediary list this way. Perhaps instead we want `allValueEntries :: Value -> (CurrencyId, TokenId, Integer) -> Bool`. Or I guess the direct way is fine too.

Comment (michaelpj) [reply plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs:324]: At least worth a comment explaining why you didn't use the existing function so I don't just refactor it back again later ;)

Comment (zliu41) [reply plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs:324]: > Aren't you pretty much exactly flattening the second value?

There's no need to flatten the _entire_ second value. We can short circuit.

Comment (zliu41) [reply plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs:324]: I guess this is the limitation with strict languages - it's much harder to fuse transformations without materializing intermediate results.

Comment (michaelpj) [reply plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs:326]: > There's no need to flatten the entire second value. We can short circuit.

Good point. Comment!

**Pull request merged**